metrics/metrics.go

```diff
@@ -0,0 +1,37 @@
+// Package metrics provides an extensible framework to instrument your
+// application. All metrics are safe for concurrent use. Considerable design
+// influence has been taken from https://github.com/codahale/metrics and
+// https://prometheus.io.
+package metrics
+
+// Counter is a monotonically-increasing, unsigned, 64-bit integer used to
+// capture the number of times an event has occurred. By tracking the deltas
+// between measurements of a counter over intervals of time, an aggregation
+// layer can derive rates, acceleration, etc.
+type Counter interface {
+	With(Field) Counter
+	Add(delta uint64)
+}
+
+// Gauge captures instantaneous measurements of something using signed, 64-bit
+// integers. The value does not need to be monotonic.
+type Gauge interface {
```

Comment:
Allow me to walk my statement back a bit :) — I actually rarely use gauges. When I do, they typically reflect allocation of resources.

```go
func process() {
    // ...
    something.Allocate(n)
    myGauge.Add(n)
    defer myGauge.Add(-n)
    // ...
}
```

In general the API borrows heavily from [Prometheus](http://godoc.org/github.com/prometheus/client_golang/prometheus#Gauge), and so to adapt their example:

```go
ops := getWork()
opsQueued.Add(len(ops))
for _, op := range ops {
    c <- op
}

// ...

for op := range c {
    opsQueued.Add(-1)
    // ...
}
```